src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -87,10 +89,44 @@ public bool ReverseOrder
         }
     }
 
+    // The encoding and decoding of LDAP directory controls interprets BER INTEGER values as 32-bit signed integers.
+    // Although BER INTEGERS can exceed this data type's maximum value, previous versions of DirectoryControl (and
+    // its derived classes) encoded and decoded these types by passing the "i" format specifier to BerConverter. The
+    // .NET Framework continues to do so. There is therefore historical precedent to justify limiting BER INTEGER
+    // values to this data type when using AsnDecoder and AsnWriter.
     public class DirectoryControl
     {
+        // Scratch buffer allocations with sizes which are below this threshold should be made on the stack.
+        // This is partially based on RFC1035, which specifies that a label in a domain name should be < 64 characters.
+        // If a server name is specified as an FQDN, this will be at least three labels in an AD environment - up to
+        // 192 characters. Doubling this to allow for Unicode encoding, then rounding to the nearest power of two
+        // yields 512.
+        internal const int ServerNameStackAllocationThreshold = 512;
+        // Scratch buffer allocations with sizes which are below this threshold should be made on the stack.
+        // This is based on the Active Directory schema. The largest attribute name here is msDS-FailedInteractiveLogonCountAtLastSuccessfulLogon,
+        // which is 53 characters long. This is rounded up to the nearest power of two.
+        internal const int AttributeNameStackAllocationThreshold = 64;
+
+        internal static readonly UTF8Encoding s_utf8Encoding = new(false, true);
+
         internal byte[] _directoryControlValue;
 
+        [ThreadStatic]
+        private static AsnWriter t_smallWriter;
+        [ThreadStatic]
+        private static AsnWriter t_mediumWriter;
+        [ThreadStatic]
+        private static AsnWriter t_largeWriter;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static AsnWriter GetWriter(int expectedSize)
+            => expectedSize switch
+            {
+                > 0 and <= 32 => t_smallWriter ??= new AsnWriter(AsnEncodingRules.BER, 32),
+                > 32 and <= 128 => t_mediumWriter ??= new AsnWriter(AsnEncodingRules.BER, 128),
+                _ => t_largeWriter ??= new AsnWriter(AsnEncodingRules.BER, 256)
+            };
```

Comment:
```suggestion
        [ThreadStatic]
        private static AsnWriter t_writer;

        internal static AsnWriter GetWriter(int expectedSize)
            => t_writer ??= new AsnWriter(AsnEncodingRules.BER);
```

Basically, once two of these get activated, or one of them grows, you're better off just having one.

Only ever needed to write a 5 byte value `30 03 02 01 00` (`SEQUENCE(INTEGER(0))`)? Sure, 1019 bytes of the heap could _maybe_ be better used; but at the end of the day it doesn't really matter, especially once it's hanging out in a thread-static.

Reply:
That's fair. It's not ideal to see the existing serialization use 300-400 bytes of memory, then to see my changes double that usage on a buffer which isn't used - but the existing serialization has unmanaged memory usage behind it too, and by the time we've sent three controls the PR's memory usage should come out ahead.

I understand, and I'll change this (with the knock-on implication: those size calculations will also vanish.) Thanks for the explanations.